apps/editing-toolkit/editing-toolkit-plugin/tutorials/class-wpcom-tutorials.php

```diff
@@ -0,0 +1,227 @@
+		return $tutorial;
+	}
+
+	/**
+	 * Retrives all registered Tutorial IDs
+	 *
+	 * @return array list of registered Tutorial IDs.
+	 */
+	public function get_registered_ids() {
+		return array_keys( $this->registry );
+	}
+
+	/**
+	 * Gets registered tasks for a Tutorial, with stored completion status.
+	 *
+	 * @param array $tutorial Tutorial object.
+	 * @return array          Tasks for Tutorial, with task completion status.
+	 */
+	private function get_tasks( $tutorial ) {
+		$tasks_with_statuses = array();
+		$statuses            = $this->get_statuses( $tutorial['id'] );
+
+		foreach ( $tutorial['tasks'] as $task ) {
+			$task['status']        = isset( $statuses[ $task['id'] ] ) ? $statuses[ $task['id'] ] : 'pending';
+			$tasks_with_statuses[] = $task;
+		}
+
+		return $tasks_with_statuses;
+	}
+
+	/**
+	 * Gets the completion statuses for a specific Tutorial.
+	 *
+	 * @param string $id Tutorial ID.
+	 * @return array     Completion statuses for Tutorial ID.
+	 */
+	public function get_statuses( $id ) {
+		$statuses = $this->load_persisted_statuses();
+		return isset( $statuses[ $id ] ) ? $statuses[ $id ] : array();
+	}
+
+	/**
+	 * Loads all Tutorial statuses from the database.
+	 *
+	 * @return array All Tutorial completion statuses.
+	 */
+	private function load_persisted_statuses() {
+		if ( function_exists( 'get_user_attribute' ) ) {
+			return get_user_attribute( get_current_user_id(), 'wpcom_tutorials_status' );
+		}
+		return (array) get_user_meta( get_current_user_id(), 'wpcom_tutorials_status', true );
+	}
+
+	/**
+	 * Marks a task with a status.
+	 *
+	 * @param string $tutorial_id Tutorial ID.
+	 * @param string $task_id     Task ID.
+	 * @param string $status      Status for the task. Allowed: complete|skipped|pending.
+	 * @return bool               True if status is updated, false otherwise.
+	 */
+	public function mark_task( $tutorial_id, $task_id, $status ) {
+		if ( ! $this->validate_status( $status ) ) {
+			return false;
+		}
+
+		$statuses = $this->load_persisted_statuses();
+		if ( ! isset( $statuses[ $tutorial_id ] ) ) {
+			$statuses[ $tutorial_id ] = array();
+		}
+		// pending is the natural state with no stored status.
+		if ( 'pending' === $status ) {
+			if ( isset( $statuses[ $tutorial_id ][ $task_id ] ) ) {
+				unset( $statuses[ $tutorial_id ][ $task_id ] );
+			}
+		} else {
+			$statuses[ $tutorial_id ][ $task_id ] = $status;
+		}
+
+		return $this->persist_statuses( $statuses );
+	}
+
+	/**
+	 * Persist statuses to the DB, using the platform-appropriate tech
+	 *
+	 * @param array $statuses All statuses to persist.
+	 * @return bool           True if persisting succeeded, otherwise false.
+	 */
+	public function persist_statuses( $statuses ) {
+		if ( function_exists( 'update_user_attribute' ) ) {
+			return update_user_attribute( get_current_user_id(), 'wpcom_tutorials_status', $statuses );
+		}
+		return update_user_meta( get_current_user_id(), 'wpcom_tutorials_status', $statuses );
+	}
+
```

Comment:
Just for my own understanding it looks like we are saving this as 'wpcom_tutorials_status' of the form:
```
{
    tutorial_1_id:    {
                                 task_1_id: task_1_status,
                                 ...
                             },
    ...
}
```

and retrieving from `get` gives us the form
```
{
    id: tutorial_id
    title:  tutorial_title,
    tasks:  [
                      {
                           id: task_1_id,
                           title: task_1_title,
                           status: task_1_status,
                       },
                      ...
   ]
}
```
does that seem accurate?

Reply:
Yes that's accurate. The `pending` status is never saved and a task with a previously-set status will have that removed if we're updating to it.

Reply:
Thanks for confirming, I think that makes sense to me!

My first thought was that perhaps tasks should be an associative array, so we can easily look up status by task_id on the front-end (`task_status = tutorial['tasks']['task_id']['status']` kind of thing). 

But thinking about it further Im not sure that would really buy us anything at the scale were are talking about for the amount of tasks that will be present (`task_status = tutorial['tasks'].find( task => task.id == task_id)?.['status'] seems fine at our scale). Any thoughts there?